pkg/operator/controller/route-deprecated/controller.go

```diff
@@ -0,0 +1,227 @@
+package routedeprecated
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	operatorcontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	"github.com/google/go-cmp/cmp"
+	"golang.org/x/time/rate"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	controllerName         = "route_deprecated_controller"
+	routeConfigAck415      = "ack-4.15-route-config-not-supported-in-4.16"
+	routeConfigAck415Msg   = "A route in this cluster contains a configuration that isn't supported in 4.16. Please review each route's deprecated status condition for more information."
+	adminGateConfigMapName = "admin-gates"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route deprecated controller. This is the controller
+// that handles interpreting the route deprecated status into an admin gate
+// which will block upgrades until an admin acks it or resolves the deprecation.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	operatorCache := mgr.GetCache()
+	reconciler := &reconciler{
+		cache:     config.RouteCache,
+		namespace: config.Namespace,
+		client:    mgr.GetClient(),
+	}
+
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler: reconciler,
+		RateLimiter: workqueue.NewMaxOfRateLimiter(
+			// Rate-limit to 1 update every 5 seconds per
+			// reconcile to avoid burning CPU if object
+			// updates are frequent.
+			workqueue.NewItemExponentialFailureRateLimiter(5*time.Second, 30*time.Second),
+			// 10 qps, 100 bucket size, same as DefaultControllerRateLimiter().
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		),
+	})
+
+	// Index routes over the deprecated condition so that we can
+	// efficiently find deprecated routes.
+	if err := reconciler.cache.IndexField(context.Background(), &routev1.Route{}, "deprecatedRoute", client.IndexerFunc(func(o client.Object) []string {
+		if cond := getRouteDeprecated(o.(*routev1.Route)); cond != nil {
```

Comment:
Please name the function slightly more descriptively:
```suggestion
		if cond := getRouteDeprecatedStatusCondition(o.(*routev1.Route)); cond != nil {
```

Reply:
Done.